File: src/com/chinex/boroja/generics/inout/Generics.java
package com.chinex.boroja.generics.inout;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

//public class Generics<T> {}
public class Generics {
    /**public void doSomething(T t) {
        System.out.println(t);
    } */
    public void makeNoises(List<? extends Animal> animalList) {
        for (Animal a: animalList) {
            a.makeNoise();
        }
    }
//covariant in Java
    public void addToList(List<? super Mammal> animalList) {
        Mammal mammal1 = new Mammal();
        Animal animal1 = new Animal();
        Dog dog1 = new Dog();

        animalList.add(animal1);
        animalList.add(mammal1);
        animalList.add(dog1);
    }
    public void getFromList(List<? extends Dog> animalList) {
        Dog animal = animalList.get(0);
    }
    public static void main(String[] args) {
        //Generics<Dog> dogGenerics = new Generics<>();
        Mammal mammal1 = new Mammal();
        Mammal mammal2 = new Mammal();
        List<Mammal> mammals = new ArrayList<>();
        mammals.add(mammal1);
        mammals.add(mammal2);

        Animal animal1 = new Animal();
        Animal animal2 = new Animal();
        List<Animal> animals = new ArrayList<>();
        animals.add(animal1);
        animals.add(animal2);

        Dog dog1 = new Dog();
        Dog dog2 = new Dog();
        List<Dog> dogs = new ArrayList<>();
        dogs.add(dog1);
        dogs.add(dog2);

        Generics g = new Generics();
        g.makeNoises(animals);

        Comparator<Dog> dogComparator = new Comparator<Dog>() {
            @Override
            public int compare(Dog o1, Dog o2) {
                return 0;
            }
        };
        Comparator<Animal> animalComparator = new Comparator<Animal>() {
            @Override
            public int compare(Animal o1, Animal o2) {
                return 0;
            }
        };
        dogs.sort(animalComparator);

        /**Mammal mammal = new Mammal();
        Animal animal = new Animal();
        Dog dog = new Dog();
        dogGenerics.doSomething(dog); */
    }
}

File: src/com/chinex/boroja/generics/kotlin/Generics.kt
package com.chinex.boroja.generics.kotlin

import com.chinex.boroja.generics.inout.Animal
import com.chinex.boroja.generics.inout.Dog
import com.chinex.boroja.generics.inout.Generics
import com.chinex.boroja.generics.inout.Mammal

/**class Generics<T> {
    fun doSomething(t: T) {
        println(t)
    }
}*/
class Generics {

    fun makeNoises(animalList: List<Animal>) {
        for (a in animalList) {
            a.makeNoise()
        }
    }

    fun addToList(animalList: MutableList<in Mammal?>) {
        val mammal1 = Mammal()
        val animal1 = Animal()
        val dog1 = Dog()
        animalList.add(animal1)
        animalList.add(mammal1)
        animalList.add(dog1)
    }

    fun getFromList(animalList: List<out Animal>) {
        val animal = animalList[0]
    }
    //how to copy
    fun copy(src: List<out Animal>, dest: MutableList<in Animal>) {
          for (animal in src) {
            dest.add(animal)
        }
    }

}

fun main(args: Array<String>) {

    //Generics<Dog> dogGenerics = new Generics<>();
    val mammal1 = Mammal()
    val mammal2 = Mammal()
    val mammals: List<Mammal> = listOf(mammal1, mammal2)

    val animal1 = Animal()
    val animal2 = Animal()
    val animals: List<Animal> = listOf(animal1, animal2)

    val dog1 = Dog()
    val dog2 = Dog()
    val dogs: List<Dog> = listOf(dog1, dog2)

    val g = com.chinex.boroja.generics.kotlin.Generics()
    g.makeNoises(dogs)

    val copy = com.chinex.boroja.generics.kotlin.Generics()
    copy.copy(dogs, mutableListOf())

    val dogComparator: Comparator<Dog> =
        Comparator<Dog> { o1, o2 -> 0 }
    val animalComparator: Comparator<Animal> =
        Comparator<Animal> { o1, o2 -> 0 }
    //dogs.sortedWith(dogComparator)

    /**val dogGenerics: Generics<Dog> = Generics<Dog>()
    val mammal = Mammal()
    val animal = TestEdible()
    val dog = Dog()
    dogGenerics.doSomething(dog)*/
}
